### PR TITLE
Print string for comparison operands only once when it is same for both

### PR DESCRIPTION
**Pull request by N-Dekker**
When both operands of a failed comparison get the very same string representation from `FormatForComparisonFailureMessage`, it is not necessary to print it twice. For example:

    auto x = std::numeric_limits<int64_t>::lowest();
    auto y = std::numeric_limits<int64_t>::min();
    ASSERT_NE(x, y);

Output before this commit:

> Expected: (x) != (y), actual: -9223372036854775808 vs -9223372036854775808

Output after this commit:

> Expected: (x) != (y), actual: -9223372036854775808 (same for both operands)

### CODE REVIEW

**Comment by N-Dekker** [comment]
@kuzkry Hi Krystian! As you reviewed my previous pull request (#3184) so carefully, would you also like to look into this one?

**Review by kuzkry — APPROVED** [review]
Hey, sure. Since you're doing 2 things here I would suggest to create 2 separate PRs. Have a look at these pull requests for my GoogleTest fork:
https://github.com/kuzkry/googletest/pull/2
https://github.com/kuzkry/googletest/pull/3

What do you think? I think this separation makes it crystal clear.

But in general, I like this change.

**Comment by N-Dekker** [comment]
> Since you're doing 2 things here I would suggest to create 2 separate PRs.

Nice suggestion, but what would you think of 2 separate commits within a single PR? Otherwise, the second PR (avoiding to print the same string twice) would depend on the first one (in order to get the right string for any operand type).

**Comment by kuzkry** [comment]
> (...) what would you think of 2 separate commits within a single PR?

In theory, if my understanding of [the 2nd rule of contribution rules](https://github.com/google/googletest/blob/ca4b7c9ff4d8a5c37ac51795b03ffe934958aeff/CONTRIBUTING.md#contributing-a-patch) is correct, then yes, you should. I think those are really two separate things that happen to touch the same code :) Additionally, it enables maintainers to take the 1st commit and reject other(s) they dislike.

The bigger picture I see here is this:
- GitHub inability to mark PRs as dependent on others is a well-known [issue](https://github.com/isaacs/github/issues/959). People have been requesting a solution for a long time. This doesn't work nicely with the "one change, one PR" rule of GoogleTest but it is what it is. I understand your concern and what's worse, I've had times when maintainers were prematurely closing my PRs as they thought those were duplicates.
- 2 PRs x 1 commit each here is fine. Sometimes, you may find yourself on the opposite side, i.e. you'll try to deliver multiple commits under one PR as it makes no sense to separate them (it would break some logic cause it to lose meaning / be incomplete).

**Comment by N-Dekker** [comment]
>  2 PRs x 1 commit each here is fine.

@kuzkry  Honesty I'm not so sure. I would rather not present the two commits as "independent" separate PRs for the master branch (if that's what you mean). Because the first commit _"reuse CmpHelperOpFailure()"_ is intended to be merely a coding style improvement, right? While the second commit is the one that aims to change the behavior, for any type of operand (by printing _"same for both operands"_). However, when the two commits are committed in reverse order, both commits will cause behavior changes!!!

More specifically: the _"reuse CmpHelperOpFailure()"_ commit affects the `CmpHelper##op_name` functions from "gtest.cc", that are called when comparing "big integers" like `int64_t`. If the _"same for both operands"_ commit is merged onto the master before the _"reuse CmpHelperOpFailure()"_ commit, it will have the _"same for both operands"_ behavior for `int`, but not for `int64_t`. I don't think that would be desirable.

Note that I just added a (missing) unit test for `int64_t`, to clarify my point. This unit test will fail without the _"reuse CmpHelperOpFailure()"_ change. https://github.com/google/googletest/compare/0ef094adc866d565cda807bfb48941fc407fd14d..cfd65970d07c7b8d30f82fcb0c9afe0e626269ae

You see the problem? Instead of creating two independent PRs, it would still be OK to me to make this 1 PR of 2 commits... but would that possibly still get your approval as well?  😃

File: googletest/test/gtest_unittest.cc
Comment (kuzkry) [inline googletest/test/gtest_unittest.cc:4559]: Extra pair of parenthesis not needed here. Compare with the previous line.

**Comment by kuzkry** [comment]
> However, when the two commits are committed in reverse order, both commits will cause behavior changes!!!

Haha, actually yes. Obviously order would matter here but that's the case when there's dependency. Anyway, I respect you don't want complications.

> You see the problem? Instead of creating two independent PRs, it would still be OK to me to make this 1 PR of 2 commits... but would that possibly still get your approval as well? 😃

Sure. Just fix that small format issue please and it'll be perfect.

File: googletest/test/gtest_unittest.cc
Comment (N-Dekker) [reply googletest/test/gtest_unittest.cc:4559]: @kuzkry Are you sure the extra pair of parenthesis not needed here? When I remove them, VS2019 says:

> gtest_unittest.cc(4559,69): error C2220: the following warning is treated as an error
> gtest_unittest.cc(4559,69): warning C4003: not enough arguments for function-like macro invocation 'min'
> gtest_unittest.cc(4559,69): error C2589: '(': illegal token on right side of '::'
> gtest_unittest.cc(4559,69): error C2062: type 'unknown-type' unexpected
> gtest_unittest.cc(4559,69): error C2059: syntax error: ')'
> gtest_unittest.cc(4563,1): error C3536: 'min_int64': cannot be used before it is initialized

As it tries to call the `min(a,b)`macro from "minwindef.h".

Comment (kuzkry) [reply googletest/test/gtest_unittest.cc:4559]: LOL, I didn't expect that 😄 I only verified it on my Linux machine. You're right

**Review by kuzkry — APPROVED** [review]

File: googletest/test/gtest_unittest.cc
Comment (N-Dekker) [reply googletest/test/gtest_unittest.cc:4559]: @kuzkry No problem, thanks! Instead of the extra parentheses, I could have added `#define NOMINMAX`, or `#undef min`, but some people may not like those hashtags either 😸  

Thanks also for your approval  😃

**Comment by kuzkry** [comment]
Hey! FYI, I find some of this code spurious and I'm gonna attempt to remove it. Depending what gets first, your PR may conflict with #3204. Happy new year btw. :)

**Comment by N-Dekker** [comment]
> FYI, I find some of this code spurious and I'm gonna attempt to remove it. Depending what gets first, your PR may conflict with #3204.

@kuzkry Thanks for informing me about our potential conflict, Krystian! Now I just hope that mine gets merged first 😸 Happy new year anyway!  :)

**Comment by N-Dekker** [comment]
> Depending what gets first, your PR may conflict with #3204.

I guess such a (possible) conflict could be fixed by simply doing a `git rebase -i HEAD~2`, and then removing the first commit from this PR, [Reuse existing CmpHelperOpFailure() in GTEST_IMPL_CMP_HELPER_ (gtest.cc)](https://github.com/google/googletest/pull/3191/commits/724a650b5baf9fd504340d992531ad2f192eeed4), right?

**Comment by kuzkry** [comment]
That's right :) It's good you're doing notes cause some things take time to be delivered here so you could forget about this merge conflict 10 times 😄

**Comment by derekmauro** [comment]
Thanks for the pull request. If we were to do this over, I think we'd print a message similar to the one you are proposing. Unfortunately though, we have to deal with [Hyrum's Law](https://www.hyrumslaw.com) in the sense that there may be code depending on the message that has been baked in for a while. The cost of the churn here outweighs the cost of having a better message.

Sorry about this.

**Comment by N-Dekker** [comment]
@derekmauro Thanks for having reviewed my pull request. Of course, I'm very disappointed about the rejection of my proposed improvement.

> there may be code depending on the message that has been baked in for a while

Do you possibly have an example of a useful piece of code in the real world, that would be broken by improving the error message of `EXPECT_NE` and `ASSERT_NE`?

**Comment by N-Dekker** [comment]
> Unfortunately though, we have to deal with Hyrum's Law

FYI, I just informed Hyrum about his law being dealt with: https://twitter.com/Niels_Dekker/status/1350450521132453888

**Pull request closed without merging**
